shared/app/tab-bar/container.js

```diff
@@ -1,20 +1,73 @@
 // @flow
-import {createSelector, connect, type TypedState} from '../../util/container'
+import {connect, type TypedState, type Dispatch, isMobile} from '../../util/container'
 import {usernameSelector} from '../../constants/selectors'
 import TabBarRender from './index.render'
+import {chatTab, peopleTab, profileTab, type Tab} from '../../constants/tabs'
+import {navigateTo, switchTo} from '../../actions/route-tree'
+import {createShowUserProfile} from '../../actions/profile-gen'
 
-const getNavBadges = (state: TypedState) => state.notifications.get('navBadges')
+const mapStateToProps = (state: TypedState): * => ({
+  _badgeNumbers: state.notifications.get('navBadges'),
+  username: usernameSelector(state),
+})
+
+const mapDispatchToProps = (dispatch: Dispatch, {routeSelected, routePath}) => ({
+  _onTabClick: isMobile
+    ? (tab: Tab, me: ?string) => {
+        if (tab === chatTab && routeSelected === tab) {
+          dispatch(navigateTo(routePath.push(tab)))
+          return
+        }
+
+        // If we're going to the people tab, switch to the current user's
+        // profile first before switching tabs, if necessary.
+        if (tab === peopleTab) {
+          if (routeSelected === tab) {
+            // clicking on profile tab when already selected should back out to root profile page
+            dispatch(navigateTo([], [peopleTab]))
+          }
+          dispatch(switchTo([peopleTab]))
+          return
+        }
+
+        const action = routeSelected === tab ? navigateTo : switchTo
+        dispatch(action(routePath.push(tab)))
+      }
+    : (tab: Tab, me: ?string) => {
+        if (tab === chatTab && routeSelected === tab) {
+          // clicking the chat tab when already selected should do nothing.
+          return
+        }
 
-const mapStateToProps = createSelector([getNavBadges, usernameSelector], (badgeNumbers, username) => ({
-  badgeNumbers: badgeNumbers.toObject(),
```

Comment:
toObject in a mapStateToProps will always cause you to thrash whenver anything is happening